File: frontend/src/lib/billing.ts
import { api } from './api';

export type PlanCatalogItem = {
  plan: string;
  monthly_price_id: string;
  yearly_price_id: string;
  base_seats: number;
  monthly_usd: number;
  yearly_usd: number;
  formatted_monthly: string;
  formatted_yearly: string;
};

export type PlanCatalogResponse = {
  plans: PlanCatalogItem[];
  additional_seat_usd_monthly: number;
  ai_overage_usd_per_1k_tokens: number;
  self_managed: boolean;
};

export type TenantSubscriptionRead = {
  id: string;
  plan: string;
  status: string;
  seat_count: number;
  seat_quota: number;
  overage_seats: number;
  interval: string;
  current_period_start: string | null;
  current_period_end: string | null;
  cancel_at_period_end: boolean;
  canceled_at: string | null;
  trial_end: string | null;
  currency: string;
  amount_per_period: number | string;
  stripe_subscription_id: string | null;
  stripe_customer_id: string | null;
  created_at: string;
  updated_at: string;
};

export type UsageMeterRead = {
  meter_key: string;
  period_start: string;
  period_end: string;
  quantity: number | string;
  reported_to_stripe_at: string | null;
};

export type UsageResponse = {
  meters: UsageMeterRead[];
  ai_paid_usd_this_period: number;
  ai_monthly_cap_usd: number;
};

export type InvoiceRead = {
  id: string;
  number: string | null;
  status: string;
  amount_due_usd: number;
  amount_paid_usd: number;
  currency: string;
  created: number;
  period_start: number | null;
  period_end: number | null;
  hosted_invoice_url: string | null;
  invoice_pdf: string | null;
};

export type InvoiceListResponse = {
  invoices: InvoiceRead[];
  self_managed: boolean;
};

export type CheckoutIn = {
  plan: 'core' | 'edu' | 'verticals';
  interval?: 'month' | 'year';
  success_url?: string;
  cancel_url?: string;
};

export type CheckoutOut = { checkout_url: string; self_managed: boolean };
export type PortalOut = { portal_url: string; self_managed: boolean };

export type BillingActionResponse = {
  subscription: TenantSubscriptionRead | null;
  message: string;
};

const PREFIX = '/billing';

export const billingApi = {
  listPlans: () => api.get<PlanCatalogResponse>(`${PREFIX}/plans`).then((r) => r.data),
  getSubscription: () =>
    api.get<TenantSubscriptionRead | null>(`${PREFIX}/subscription`).then((r) => r.data),
  startCheckout: (body: CheckoutIn) =>
    api.post<CheckoutOut>(`${PREFIX}/checkout`, body).then((r) => r.data),
  openPortal: (returnUrl?: string) =>
    api.post<PortalOut>(`${PREFIX}/portal`, { return_url: returnUrl }).then((r) => r.data),
  cancel: () =>
    api.post<BillingActionResponse>(`${PREFIX}/cancel`, {}).then((r) => r.data),
  resume: () =>
    api.post<BillingActionResponse>(`${PREFIX}/resume`, {}).then((r) => r.data),
  usage: () => api.get<UsageResponse>(`${PREFIX}/usage`).then((r) => r.data),
  invoices: () => api.get<InvoiceListResponse>(`${PREFIX}/invoices`).then((r) => r.data),
};

File: backend/app/schemas/billing.py
"""Pydantic schemas for the billing module."""
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.common import ORMModel


class PlanCatalogItem(BaseModel):
    """One row in the public pricing catalog returned by ``GET /billing/plans``."""

    plan: str
    monthly_price_id: str
    yearly_price_id: str
    base_seats: int
    monthly_usd: float
    yearly_usd: float
    formatted_monthly: str  # e.g. "$99/mo"
    formatted_yearly: str   # e.g. "$990/yr (save 17%)"


class PlanCatalogResponse(BaseModel):
    plans: list[PlanCatalogItem]
    additional_seat_usd_monthly: float
    ai_overage_usd_per_1k_tokens: float
    self_managed: bool  # True when STRIPE_SECRET_KEY isn't set


class TenantSubscriptionRead(ORMModel):
    id: str
    plan: str
    status: str
    seat_count: int
    seat_quota: int
    overage_seats: int
    interval: str
    current_period_start: datetime | None = None
    current_period_end: datetime | None = None
    cancel_at_period_end: bool
    canceled_at: datetime | None = None
    trial_end: datetime | None = None
    currency: str
    amount_per_period: Decimal
    stripe_subscription_id: str | None = None
    stripe_customer_id: str | None = None
    created_at: datetime
    updated_at: datetime


class CheckoutIn(BaseModel):
    plan: Literal["core", "edu", "verticals"]
    interval: Literal["month", "year"] = "month"
    success_url: str | None = Field(default=None, max_length=512)
    cancel_url: str | None = Field(default=None, max_length=512)


class CheckoutOut(BaseModel):
    checkout_url: str
    self_managed: bool  # True when Stripe isn't configured — URL points at docs


class PortalIn(BaseModel):
    return_url: str | None = Field(default=None, max_length=512)


class PortalOut(BaseModel):
    portal_url: str
    self_managed: bool


class UsageMeterRead(ORMModel):
    meter_key: str
    period_start: date
    period_end: date
    quantity: Decimal
    reported_to_stripe_at: datetime | None = None


class UsageResponse(BaseModel):
    meters: list[UsageMeterRead]
    # Plain-language convenience field — sum of paid AI cost this period.
    ai_paid_usd_this_period: float = 0.0
    ai_monthly_cap_usd: float = 0.0


class InvoiceRead(BaseModel):
    """Subset of the Stripe Invoice object we surface to the UI."""

    id: str
    number: str | None = None
    status: str
    amount_due_usd: float
    amount_paid_usd: float
    currency: str
    created: int
    period_start: int | None = None
    period_end: int | None = None
    hosted_invoice_url: str | None = None
    invoice_pdf: str | None = None


class InvoiceListResponse(BaseModel):
    invoices: list[InvoiceRead]
    self_managed: bool


class BillingActionResponse(BaseModel):
    """Generic response for cancel / resume."""

    subscription: TenantSubscriptionRead | None = None
    message: str
